packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -166,120 +169,201 @@ class TimelineData {
+        if (eventAdded) _pendingEvents.remove(event);
+      }
     }
   }
 
-  void _handleAsyncInstantEvent(TraceEvent event) {
-    final String asyncUID = event.asyncUID;
+  /// Add event to an available frame in [_pendingFrames] if we can, or otherwise add it
+  /// to [_pendingEvents].
+  void _maybeAddEvent(TimelineEvent event) {
+    if (!event.isPipelineProduceFlow && !event.isPipelineConsumeFlow) {
+      // We do not care about other events.
+      return;
+    }
 
-    final TimelineEvent e = TimelineEvent(
-      event.name,
-      event.timestampMicros,
-      _getTimelineEventType(event),
-    );
-    e.endTime = event.timestampMicros + event.duration;
+    bool eventAdded = false;
 
-    final TimelineEvent parent = _asyncEvents[asyncUID];
-    if (parent != null) {
-      e.parent = parent;
+    final List<TimelineFrame> frames = _getAndSortWellFormedFrames();
+    for (TimelineFrame frame in frames) {
+      eventAdded = _maybeAddEventToFrame(event, frame);
     }
+
+    if (!eventAdded) _pendingEvents.add(event);
   }
 
-  void _handleAsyncEndEvent(TraceEvent event) {
-    final String asyncUID = event.asyncUID;
+  /// Add event `e` to frame `f` if it meets the necessary criteria.
+  ///
+  /// Returns a bool indicating whether the event was added to the frame.
+  bool _maybeAddEventToFrame(TimelineEvent e, TimelineFrame f) {
+    assert(f.wellFormed);
 
-    final TimelineEvent current = _asyncEvents[asyncUID];
-    if (current != null) {
-      current.endTime = event.timestampMicros;
-      _asyncEvents[asyncUID] = current.parent;
+    // TODO(kenzie): consider trimming VSYNC layer from pipelineProduceFlow. It
+    // can start outside of the frame's time boundaries and could pose a risk
+    // for us missing a frame.
 
-      if (_asyncEvents[asyncUID] == null) {
-        frames[event.frameId].addEvent(current);
-      }
+    // Ensure the event fits within the frame's time boundaries.
+    if (!_eventOccursWithinFrameBoundaries(e, f)) return false;
+
+    bool eventAdded = false;
+
+    if (e.isPipelineProduceFlow && f.pipelineProduceFlow == null) {
+      f.pipelineProduceFlow = e;
+      eventAdded = true;
+    } else if (e.isPipelineConsumeFlow && f.pipelineConsumeFlow == null) {
+      f.pipelineConsumeFlow = e;
+      eventAdded = true;
     }
+
+    // Adding event 'e' could mean we have completed the frame. Check if we
+    // should add the completed frame to [_frameCompleteController].
+    _maybeAddCompletedFrame(f);
+
+    return eventAdded;
   }
 
-  TimelineEventType _getTimelineEventType(TraceEvent event) {
-    // The event will only be of type 'cpu' or 'gpu' because we do not process
-    // events of any other type.
-    if (event.isCpuEvent) {
-      return TimelineEventType.cpu;
-    } else {
-      return TimelineEventType.gpu;
+  bool _eventOccursWithinFrameBoundaries(TimelineEvent e, TimelineFrame f) {
+    // Epsilon in microseconds.
+    const int epsilon = 50;
```

Comment:
why do we need an epsilon of 50 ms? That seems a bit surprising.

Reply:
Explained below. Added a todo to talk to engine team.